bioio_base/embedded_metadata.py

```diff
@@ -0,0 +1,136 @@
+# Descriptor for metadata fields.
+# -----------------------------------------------------------------------------
+T = TypeVar("T")  # Generic type variable to enable type-safe storage of metadata
+
+
+class MetadataField(Generic[T]):
+    def __init__(self, label: str, default: Optional[T] = None) -> None:
+        self.label = label
+        self.default = default
+        self.private_name: Optional[str] = None
+
+    def __set_name__(self, owner: Any, name: str) -> None:
+        self.private_name = "_" + name
+
+    @overload
+    def __get__(self, instance: None, owner: Any) -> "MetadataField[T]":
+        ...
+
+    @overload
+    def __get__(self, instance: Any, owner: Any) -> Optional[T]:
+        ...
+
+    def __get__(
+        self, instance: Any, owner: Any = None
+    ) -> Union["MetadataField[T]", Optional[T]]:
+        if instance is None:
+            return self  # When accessed on the class, return the descriptor.
+        return instance.__dict__.get(self.private_name, self.default)
+
+    def __set__(self, instance: Any, value: T) -> None:
+        instance.__dict__[self.private_name] = value
+
+
+# -----------------------------------------------------------------------------
+# EmbeddedMetadata class with definitive (closed) fields.
+# -----------------------------------------------------------------------------
+class EmbeddedMetadata:
+    # Calculated metadata fields.
+    dimensions_present: MetadataField[Optional[Sequence[str]]] = MetadataField(
+        "Dimensions Present"
+    )
+    image_size_c: MetadataField[Optional[int]] = MetadataField("Image Size C")
+    image_size_t: MetadataField[Optional[int]] = MetadataField("Image Size T")
+    image_size_x: MetadataField[Optional[int]] = MetadataField("Image Size X")
+    image_size_y: MetadataField[Optional[int]] = MetadataField("Image Size Y")
+    image_size_z: MetadataField[Optional[int]] = MetadataField("Image Size Z")
+    pixel_size_x: MetadataField[Optional[float]] = MetadataField("Pixel Size X")
+    pixel_size_y: MetadataField[Optional[float]] = MetadataField("Pixel Size Y")
+    pixel_size_z: MetadataField[Optional[float]] = MetadataField("Pixel Size Z")
+    timelapse: MetadataField[Optional[bool]] = MetadataField("Timelapse")
+
+    # Predefined metadata fields for user override.
+    objective: MetadataField[Optional[str]] = MetadataField("Objective")
+    binning: MetadataField[Optional[str]] = MetadataField("Binning")
+    column: MetadataField[Optional[str]] = MetadataField("Column")
+    imaged_by: MetadataField[Optional[str]] = MetadataField("Imaged By")
+    imaging_date: MetadataField[Optional[str]] = MetadataField("Imaging Date")
+    position_index: MetadataField[Optional[int]] = MetadataField("Position Index")
+    row: MetadataField[Optional[str]] = MetadataField("Row")
+    total_time_duration: MetadataField[Optional[str]] = MetadataField(
+        "Total Time Duration"
+    )
+    well: MetadataField[Optional[str]] = MetadataField("Well")
```

Comment:
I think we can remove `well` since we have `row` and `column` - if we are fancy we could construct `well` for them using `row` and `column`

Reply:
removed in [8d2c263](https://github.com/bioio-devs/bioio-base/pull/29/commits/8d2c263e2f17e05e292673dcacc3856be8aaa39d)